src/datasets/arrow_dataset.py

```diff
@@ -3484,14 +3484,20 @@ def delete_file(file):
             buffer = BytesIO()
             shard.to_parquet(buffer)
             uploaded_size += buffer.tell()
-            api.upload_file(
-                path_or_fileobj=buffer.getvalue(),
-                path_in_repo=path_in_repo(index),
-                repo_id=repo_id,
-                token=token,
-                repo_type="dataset",
-                revision=branch,
-                identical_ok=True,
+            _retry(
+                api.upload_file,
+                func_kwargs=dict(
+                    path_or_fileobj=buffer.getvalue(),
+                    path_in_repo=path_in_repo(index),
+                    repo_id=repo_id,
+                    token=token,
+                    repo_type="dataset",
+                    revision=branch,
+                    identical_ok=True,
+                ),
+                exceptions=HTTPError,
+                status_codes=[504],
+                max_retries=2,
```

Comment:
You can set the max_wait_time to a much bigger value than the default 2sec, since an upload operation can take some time before failing, I guess ? . Maybe 30sec ?

Reply:
@lhoestq thanks for your comment, because I don't know which values could be optimal/reasonable.

Just to be sure we understand the workflow:
- we use `requests`, which makes synchronous requests
- for each shard, we perform a synchronous POST request
- this can take long time to be finished until we receive the server response (I understand you are referring to this time), but, because of synch, the execution is blocked until we receive the server response
- it is just when we receive the server response with error status code 504, that we should retry the same POST request, but only after some sleep time
- this is where `max_wait_time` is relevant: this is the max time we will sleep since we receive the 504 error until we resend the same request

In order to decide the value of this, we should know: when a server sends a 504 error, how long does it typically take to be operative again, so that it can take requests normally?

This time is different from the time the server typically take to process one request, which is what I think you are referring in your comment.

Reply:
Oh I see your point, right !

> In order to decide this value, we should know: when a server sends a 504 error, how long does it typically take to be operative again, so that it can take requests normally?

From my experience it can take up to a minute, but I'm not sure what causes them in the first place

Reply:
> From my experience it can take up to a minute

is this in prod or staging?

Reply:
For prod. I haven't played with staging that much to experience this on staging